lib/Bitcoin/Crypto/Block.pm

```diff
@@ -21,6 +55,271 @@ has option 'previous' => (
+			
+			my $script_length = unpack_compactsize $serialized, \$tx_pos;
+			$tx_pos += $script_length; # Script
+			$tx_pos += 4;  # Sequence
+		}
+		
+		# Parse output count and outputs
+		my $output_count = unpack_compactsize $serialized, \$tx_pos;
+		for (1 .. $output_count) {
+			$tx_pos += 8;  # Value
+			
+			my $script_length = unpack_compactsize $serialized, \$tx_pos;
+			$tx_pos += $script_length; # Script
+		}
+		
+		# Parse witness data if present
+		if ($witness_flag) {
+			for (1 .. $input_count) {
+				my $input_witness_count = unpack_compactsize $serialized, \$tx_pos;
+				for (1 .. $input_witness_count) {
+					my $witness_item_length = unpack_compactsize $serialized, \$tx_pos;
+					$tx_pos += $witness_item_length;
+				}
+			}
+		}
+		
+		# Locktime (4 bytes)
+		$tx_pos += 4;
+		
+		# Extract and parse transaction
+		my $tx_length = $tx_pos - $tx_start;
+		my $tx_data = substr($serialized, $tx_start, $tx_length);
+		my $tx = Bitcoin::Crypto::Transaction->from_serialized($tx_data);
+		
+		$pos = $tx_pos;
+		push @transactions, $tx;
+	}
+
+	Bitcoin::Crypto::Exception::Block->raise(
+		'serialized block data is corrupted'
+	) if $pos != length $serialized;
+
+	my $block_args = {
+		version => $version,
+		prev_block_hash => $prev_block_hash,
+		merkle_root => $merkle_root,
+		timestamp => $timestamp,
+		bits => $bits,
+		nonce => $nonce,
+	};
+	
+	# Add height if provided
+	$block_args->{height} = $height if defined $height;
+
+	my $block = $class->new($block_args);
+	@{$block->transactions} = @transactions;
+
+	return $block;
+}
+
+signature_for get_hash => (
+	method => Object,
+	positional => [],
+);
+
+sub get_hash
+{
+	my ($self) = @_;
+	return scalar reverse hash256($self->get_header);
+}
+
+signature_for get_header => (
+	method => Object,
+	positional => [],
+);
+
+sub get_header
+{
+	my ($self) = @_;
+
+	# Return just the 80-byte block header
+	my $header = '';
+	$header .= pack 'V', $self->version;
+	$header .= scalar reverse $self->prev_block_hash;
+	$header .= scalar reverse $self->merkle_root;
+	$header .= pack 'V', $self->timestamp;
+	$header .= pack 'V', $self->bits;
+	$header .= pack 'V', $self->nonce;
+
+	return $header;
+}
+
+signature_for calculate_merkle_root => (
+	method => Object,
+	positional => [],
+);
+
+sub calculate_merkle_root
+{
+	my ($self) = @_;
+
+	my @tx_hashes = map { scalar reverse $_->get_hash } @{$self->transactions};
+
+	Bitcoin::Crypto::Exception::Block->raise(
+		'cannot calculate merkle root for empty block'
+	) unless @tx_hashes;
+
+	# Build merkle tree
+	while (@tx_hashes > 1) {
+		my @next_level;
+
+		for (my $i = 0; $i < @tx_hashes; $i += 2) {
+			my $left = $tx_hashes[$i];
+			my $right = $i + 1 < @tx_hashes ? $tx_hashes[$i + 1] : $left;
+
+			# Concatenate and double hash
+			push @next_level, hash256($left . $right);
+		}
+
+		@tx_hashes = @next_level;
+	}
+
+	$self->{merkle_root} = scalar reverse $tx_hashes[0];  # set merkle_root
+
+	return $self->merkle_root;
+}
+
```

Comment:
This should probably be a (lazy) default value builder for `merkle_root` field. Modifying transaction in a way that changes its merkle root should also clear `merkle_root` field, so that it will be automatically re-calculated next time the field is accessed.